File: frontend/src/room/actions.ts
/**
 * WHERE AN OFFER GOES, AND WHAT IT SAYS IT COSTS.
 *
 * Pure. The loop validated every offer (agent/actions.py) and derived what each
 * act costs from metrics.yaml; this module only decides WHERE each one is drawn
 * and what the button reads.
 *
 * THE RULE IS ONE SENTENCE: a targeted offer is drawn on its row, and every
 * other one at the foot. It is the Netflix "Because you watched" shape — the
 * suggestion sits on the thing it is about, so it needs no sentence naming
 * which thing it means, and it carries its reason, so a person can disagree
 * with it rather than only obey or ignore it.
 *
 * WHAT THIS MODULE MUST NEVER DO IS WORK OUT A COST. `costs` arrived on the
 * frame because the server read it out of the definitions at the moment the
 * offer was made and stored it on the answer post. Deriving it again here
 * would be a second opinion about a fact, and the two could disagree — a
 * reopened thread would then show a different promise from the one the person
 * was actually shown. It is drawn exactly as it is given.
 */
import type { ActionOffer } from '../types/bob';
import type { BoardObject } from './board';
import { callOf, rowsOf, subjectOf, type AnswerTurn } from './data';
import { markFor, type Mark } from './catalogue';

/** Every offer this turn made, or none. */
export function offersOf(turn: { actions?: ActionOffer[] } | null | undefined): ActionOffer[] {
  return Array.isArray(turn?.actions) ? turn!.actions! : [];
}

/**
 * The offers that belong to one ROW of one read.
 *
 * Matched on the read's seq AND on the target, because a seq alone is not a
 * row: two objects can draw the same read, and an offer about Magnolia belongs
 * on Magnolia's row in both of them and on nobody else's.
 *
 * The comparison is the one every other subject match in the room uses —
 * trimmed, case-folded — so an offer and the row it names cannot fail to meet
 * over a capital letter.
 */
export function onRow(offers: ActionOffer[], seq: number | null | undefined,
                      subject: string | null | undefined): ActionOffer[] {
  if (seq === null || seq === undefined || !subject) return [];
  const want = subject.trim().toLowerCase();
  return offers.filter(
    (a) => a.seq === seq && typeof a.target === 'string'
      && a.target.trim().toLowerCase() === want,
  );
}

/**
 * THE MARKS THAT DRAW A NAMED ROW, and so have somewhere to put an offer.
 *
 * `line` is missing on purpose: its rows are positions on an ordered field —
 * days, weeks, hours — and a button on one of them would be an offer about a
 * Tuesday. `figure` is here because its one row IS the subject.
 */
const NAMES_ITS_ROWS = new Set<Mark>(['figure', 'dumbbell', 'ranked', 'contributors', 'table']);

/**
 * WHICH OFFER GOES WHERE, decided ONCE for the whole screen.
 *
 * The marks could each filter the offers themselves, and then nothing would
 * know what was left over — an offer on a row the board happened to draw as a
 * line chart would vanish, and vanishing is the one thing a suggestion may not
 * do. So the decision is made here, in one pass over the board the person is
 * actually looking at, and both the marks and the foot read the answer.
 *
 * Only the NEWEST turn's objects can carry an offer: the offers came with that
 * turn, and putting one on an object folded up from three questions ago would
 * be an answer to a question nobody is looking at.
 */
export function placement(
  offers: ActionOffer[], board: BoardObject[], answers: AnswerTurn[],
): { onRows: Map<string, ActionOffer[]>; foot: ActionOffer[] } {
  const onRows = new Map<string, ActionOffer[]>();
  const taken = new Set<string>();
  const newest = answers.length - 1;
  const targeted = offers.filter((a) => typeof a.target === 'string' && a.target.trim());

  for (const o of board) {
    if (o.turn !== newest || o.seq === undefined || o.seq === null) continue;
    const turn = answers[o.turn];
    if (!turn) continue;
    const rows = rowsOf(callOf(turn, o.seq));
    if (!rows.length) continue;
    if (!NAMES_ITS_ROWS.has(markFor(o, rows))) continue;
    const names = new Set(rows.map((r) => subjectOf(r)?.trim().toLowerCase()).filter(Boolean));
    // ONE OFFER, ONE PLACE. Two objects may draw the same read — a dumbbell
    // and a ranked list of the same seven shops is an ordinary board — and an
    // offer that landed on both would be the same suggestion twice, which
    // reads as two. The board is already in reading order, so the first object
    // to be able to carry it is the one nearest the top, and it wins.
    const mine = targeted.filter(
      (a) => a.seq === o.seq && !taken.has(key(a))
        && names.has(a.target!.trim().toLowerCase()),
    );
    if (!mine.length) continue;
    onRows.set(o.key, mine);
    for (const a of mine) taken.add(key(a));
  }
  return { onRows, foot: offers.filter((a) => !taken.has(key(a))) };
}

/** One offer's identity, for telling a placed one from an unplaced one. */
export function key(a: ActionOffer): string {
  return `${a.act}|${a.seq}|${(a.target ?? '').trim().toLowerCase()}|${a.argument ?? ''}`;
}

/**
 * WHAT THE BUTTON READS: the act in the person's words, then what it costs.
 *
 * The verb is the surface's own vocabulary and not the model's — "why" is the
 * word on the tile's own button already, and two words for one act is how a
 * reader ends up thinking they are two different things.
 */
const SAYS: Record<string, string> = {
  why: 'why',
  open: 'open',
  replay: 'run it again',
};

export function says(a: ActionOffer): string {
  return SAYS[a.act] ?? a.act;
}

/**
 * The cost, as the definitions worded it, and nothing added.
 *
 * An offer whose act declared no cost draws none rather than a guess: an empty
 * label is honest and "instant" would not be.
 */
export function cost(a: ActionOffer): string | null {
  const said = (a.costs ?? '').trim();
  return said ? said : null;
}

File: agent/actions.py
"""
WHAT TO DO ABOUT IT, ON THE ROW IT IS ABOUT.

The third statement `compose` carries (P2.d, 2026-09-15), beside the blocks and
the reading. An action is not a tile and not a recommendation: it is a thing
the SURFACE can already do, pointed at one row, carrying Bob's own few words
for why that row and not another.

WHAT CAME BEFORE, AND WHY THIS IS NOT IT. `action` was a field on a
`recommendation` block until P1.f retired it. That tile said "Order Aji Mix" in
a box beside the figures it was recommending, at the same weight as them, and
it lost: every answer has a `next` sentence and few answers had the tile. The
difference here is placement and cause. A suggestion that sits ON the row it is
about needs no sentence explaining which row it means, and a suggestion that
says WHY is a suggestion a person can disagree with.

THE MODEL SUPPLIES THREE THINGS AND DERIVES NOTHING:

    {"act": "why", "seq": 2, "target": "Magnolia",
     "reason": "the only shop that fell while takings rose"}

  `act`      one of the acts metrics.yaml declares, and the surface performs.
             Nothing may be suggested that the room cannot do when it is
             tapped — an offer that opens nothing is a sentence dressed as a
             control.
  `seq`      a read that returned this turn. The same anchor a block uses, and
             for the same reason: it is what the target is checked against, and
             it is how the renderer knows which mark the offer belongs inside.
  `target`   a value a row of that read carries, or a value the read was
             scoped to — `compose._backs`, the identical test a block's
             subject passes. OPTIONAL: an action about the answer rather than
             about a row names none and is drawn at the foot.
  `reason`   Bob's words. AN ANNOTATION under CLAUDE.md's bound, so it may
             point at rows and characterise them and may never name a number.

WHAT IT NEVER SUPPLIES IS THE COST. "~1s" and "a turn" are facts about this
machine, not about the shops, and a model writing them could be wrong about
them on a slow day. Each act declares its own in metrics.yaml and the renderer
draws what it finds. A person deciding whether to tap something is owed the
difference between a second and a conversation.

WHICH ACTS THERE ARE IS THE YAML'S BUSINESS, and it currently declares two:
`why`, which asks Bob and costs a turn, and `open`, which opens the object
and costs about a second. `replay` is written down there as the act that is NOT
offered, with the reason — it would need a value as well as an argument, and
the preset values live in the renderer rather than in the definitions.

A REFUSED ACTION IS DROPPED AND THE REST STAND, exactly as a refused block is
dropped and the rest of the composition draws. Nothing here costs a round trip:
there is nothing for the model to fix that would change a figure.
"""

from __future__ import annotations

from typing import Any, Mapping, Optional

from agent import reading as _reading


class Rejected(ValueError):
    """One action refused, with a reason a person could act on."""


def spec(defs: Mapping[str, Any]) -> Mapping[str, Any]:
    return ((defs.get("composition") or {}).get("actions") or {})


def acts(defs: Mapping[str, Any]) -> Mapping[str, Any]:
    return spec(defs).get("acts") or {}


def max_actions(defs: Mapping[str, Any]) -> int:
    return int(spec(defs).get("max") or 3)


def _reason(text: Any, defs: Mapping[str, Any]) -> str:
    """
    THE WHY, HELD EXACTLY WHERE A BLOCK'S CLAIM IS HELD.

    Same rule, same reason: this is an annotation over rows that are drawn with
    their own receipts beside them. A digit here would be a figure stated above
    one that has a source, by the one channel on the turn nobody can click.
    """
    bound = spec(defs).get("reason") or {}
    if not isinstance(text, str) or not text.strip():
        raise Rejected(
            "an action says WHY this row — without it, it is a button with no "
            "argument and the person has to guess what you saw"
        )
    said = " ".join(text.split())
    longest = int(bound.get("max_length") or 70)
    if bound.get("no_digits", True) and any(ch.isdigit() for ch in said):
        raise Rejected(
            "a reason carries no digits — it sits over rows that draw their "
            "own figures (metrics.yaml composition.actions.reason)"
        )
    try:
        return _reading.over_length("reason", said, {**bound, "max_length": longest},
                                    70, None, "metrics.yaml composition.actions.reason")
    except _reading.Rejected:
        raise Rejected(
            f"a reason is at most {longest} characters — it is the few words "
            f"beside a button, not the reading"
        ) from None


def _shorten(value: Any) -> str:
    text = value if isinstance(value, str) else repr(value)
    return text[:120]


def validate(
    submitted: Any,
    calls: Mapping[int, Mapping[str, Any]],
    defs: Mapping[str, Any],
) -> tuple[list[dict], list[dict]]:
    """
    Split submitted actions into the ones that may be drawn and the ones that
    may not, each refusal with a reason.

    `calls` is this turn's reads by seq — the same mapping the composition is
    validated against, so an action and a block cannot disagree about what a
    read returned.
    """
    from agent import compose as _compose

    accepted: list[dict] = []
    rejected: list[dict] = []
    if submitted is None:
        return accepted, rejected
    if not isinstance(submitted, list):
        return accepted, [{"action": _shorten(submitted), "target": None,
                           "reason": "actions are a list of objects"}]

    known = acts(defs)
    arguments = set((defs.get("composition") or {}).get("control_arguments") or [])
    cap = max_actions(defs)
    seen: set[tuple] = set()

    for item in submitted:
        try:
            if not isinstance(item, Mapping):
                raise Rejected("an action is an object naming an act, a read and a reason")
            act = item.get("act")
            if act not in known:
                raise Rejected(
                    f"an act is one of {', '.join(sorted(known))} — what the "
                    f"surface can do when it is tapped "
                    f"(metrics.yaml composition.actions.acts)"
                )
            declared: Mapping[str, Any] = known[act]

            seq = item.get("seq")
            if not isinstance(seq, int) or isinstance(seq, bool):
                raise Rejected("an action names the seq of a read that returned this turn")
            call = calls.get(seq)
            if call is None:
                raise Rejected(f"read {seq} did not run, or did not return")

            target: Optional[str] = None
            raw = item.get("target")
            if isinstance(raw, str) and raw.strip():
                if not _compose._backs(call, raw):
                    raise Rejected(
                        f"read {seq} has no row for {raw.strip()!r} — an action "
                        f"sits on a row, and this one would sit on nothing"
                    )
                target = raw.strip()
            elif raw is not None and not isinstance(raw, str):
                raise Rejected("a target is the value of a row, as a word")

            # NO ACT TAKES AN ARGUMENT TODAY, and the refusal is kept rather
            # than the branch deleted: `replay` is written down in the yaml as
            # the act that would, and the day it comes back this is where it
            # is checked. An argument arriving now is the model reaching for
            # an act that is not offered.
            argument = None
            if declared.get("needs_argument"):
                argument = item.get("argument")
                if argument not in arguments:
                    raise Rejected(
                        f"a {act} changes one of {', '.join(sorted(arguments))} "
                        f"— scope, never a threshold "
                        f"(metrics.yaml composition.control_arguments)"
                    )
            elif item.get("argument") is not None:
                raise Rejected(f"a {act} changes no argument")

            reason = _reason(item.get("reason"), defs)

            # THE SAME OFFER TWICE IS ONE OFFER. Two rounds of composing in a
            # turn is normal, and the second one repeating the first would
            # otherwise draw the button twice on the same row.
            fingerprint = (act, seq, (target or "").lower(), argument)
            if fingerprint in seen:
                raise Rejected("that is the same action on the same row, already offered")
            seen.add(fingerprint)

            if len(accepted) >= cap:
                raise Rejected(
                    f"at most {cap} actions — past that the person is choosing "
                    f"between suggestions instead of reading the figures"
                )

            drawn = {
                "act": act,
                "seq": seq,
                "target": target,
                "reason": reason,
                # DERIVED, NEVER SUBMITTED. The schema has no field for either,
                # so neither can arrive from the model.
                "costs": str(declared.get("costs") or ""),
                "model_turn": bool(declared.get("model_turn")),
            }
            if argument is not None:
                drawn["argument"] = argument
            accepted.append(drawn)
        except Rejected as exc:
            rejected.append({
                "action": _shorten(item.get("act") if isinstance(item, Mapping) else item),
                "target": (item.get("target") if isinstance(item, Mapping) else None),
                "reason": str(exc),
            })
    return accepted, rejected
